### PR TITLE
resources: relayoutAll should invalidate premeasurements

### PR DESCRIPTION
**Pull request by samouri**
**summary**
Context: `b/170538870`, `b/169596105`, and maybe but unlikely this one as well `b/150231440`.

If the viewport size changes after an InOb premeasurement, then we need to invalidate the premeasurement as a viewport size change is likely to change the size of elements. This is especially true for elements that expand to the space given (e.g. fixed in only one dimension, or layout=container).

---

_bigger picture_: I didn't do an exhaustive check of whether there are more events which should invalidate a premeasure. two ways we could design the architecture to be more resilient are:
1. don't have a delay between when measurement is taken and applied. this is done today so that the diff between experiement on and off is reduced. after removing the experiment we can work to remove this racy time window
2. stop relying on our measurements for determining whether to layout. only rely on intersects + ! display:none

### CODE REVIEW

File: src/service/resources-impl.js [hunk 1]
```diff
@@ -264,6 +264,10 @@ export class ResourcesImpl {
       }
       // With IntersectionObserver, we only need to handle viewport resize.
       if (this.relayoutAll_ || !this.intersectionObserver_) {
+        // Unfortunately, a viewport size change invalidates all premeasurements.
+        if (this.intersectionObserver_) {
+          this.resources_.forEach((r) => r.invalidatePremeasurement());
```
Comment (kristoferbaxter) [inline src/service/resources-impl.js:269]: Nit: Possible to use a more descriptive name for each member?

`r` => `resource`

File: src/web-worker/amp-worker.js [hunk 2]
```diff
@@ -84,6 +84,7 @@ class AmpWorker {
     // Use RTV to make sure we fetch prod/canary/experiment correctly.
     const useLocal = getMode().localDev || getMode().test;
     const useRtvVersion = !useLocal;
+
```
Comment (kristoferbaxter) [inline src/web-worker/amp-worker.js:87]: Nit: Seems unrelated.

**Review by kristoferbaxter — APPROVED** [review]

File: src/web-worker/amp-worker.js (hunk 2 above)
Comment (samouri) [reply src/web-worker/amp-worker.js]: yep! I'll reverting

File: src/service/resources-impl.js (hunk 1 above)
Comment (samouri) [reply src/service/resources-impl.js]: done

File: src/service/resources-impl.js [hunk 3]
```diff
@@ -264,6 +264,12 @@ export class ResourcesImpl {
       }
       // With IntersectionObserver, we only need to handle viewport resize.
       if (this.relayoutAll_ || !this.intersectionObserver_) {
+        // Unfortunately, a viewport size change invalidates all premeasurements.
+        if (this.intersectionObserver_) {
```
Comment (jridgewell) [inline src/service/resources-impl.js:268]: Nit: I would move this out of the outer conditional, and do: `this.relayoutAll_ && this.intersectionObserver_`

**Review by jridgewell — APPROVED** [review]

File: src/service/resources-impl.js (hunk 3 above)
Comment (samouri) [reply src/service/resources-impl.js]: done!

**Comment by choumx** [comment]
Great find!!
